
Copyright: not significant
Upstream: committed

--- grub-0.95+cvs20040624.old/util/mkbimage	2004-04-29 21:45:11.000000000 +0200
+++ grub-0.95+cvs20040624/util/mkbimage	2004-07-21 15:47:15.000000000 +0200
@@ -232,10 +232,10 @@
 
 if [ ! "$image_type" ]; then
 	image_type=hd;
-elif [ "$image_type" != "1.20" -a "$image_type" != "1.44" \
-  -a "$image_type" != "1.60" -a "$image_type" != "1.68" \
-  -a "$image_type" != "2.88" -a "$image_type" != "1.74" \
-  -a "$image_type" != "hd" -a "$image_type" != "1.60" ] ; then
+elif [ "$image_type" != "1.20" ] && [ "$image_type" != "1.44" ] \
+  && [ "$image_type" != "1.60" ] && [ "$image_type" != "1.68" ] \
+  && [ "$image_type" != "2.88" ] && [ "$image_type" != "1.74" ] \
+  && [ "$image_type" != "hd" ] && [ "$image_type" != "1.60" ] ; then
   error wrong_type ;
 fi
 
